### PR TITLE
document native database upserts 

### LINKED ISSUE #3921: Document native upserts
**Ready to merge when 4.6 goes live**

SME: Garren

Info from Garren: https://github.com/prisma/docs/issues/3952

He recommends two additions to [prisma.io/docs/reference/api-reference/prisma-client-reference#upsert](https://www.prisma.io/docs/reference/api-reference/prisma-client-reference#upsert):

Change 2:
- Document the new feature: adding support for Native Upsert support

### PR DESCRIPTION
**Pull request by andrew-walford-prisma**
**Do not merge before 4.6 goes live**

Add a new doc to cover native upserts

### CODE REVIEW

File: content/200-concepts/100-components/02-prisma-client/260-native-upserts.mdx [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+</TopBlock>
+
+## Criteria for native upserts
+
+### Prerequisites
+
+Prisma Client can initiate native upserts if your stack meets the following criteria:
+
+- You use Prisma version 4.6.0 or later
+- Your application uses a CockroachDB, PostgreSQL, or SQLite database
```
Comment (garrensmith) [inline content/200-concepts/100-components/02-prisma-client/260-native-upserts.mdx]: Do we refer to databases as a datasource or database? I can never remember

File: content/200-concepts/100-components/02-prisma-client/260-native-upserts.mdx [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+### <inlinecode>upsert</inlinecode> query criteria
+
+Prisma Client initiates a native upsert if your `upsert` query meets the following criteria:
```
Comment (garrensmith) [inline content/200-concepts/100-components/02-prisma-client/260-native-upserts.mdx]: I would maybe change initiates to `uses the database native upsert`

File: content/200-concepts/100-components/02-prisma-client/260-native-upserts.mdx [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+- There are no nested queries in the `upsert`'s `create` and `update` [options](/reference/api-reference/prisma-client-reference/#options-7)
+- The query updates only one model
+- There is only one unique field in the `upsert`'s `where` option
+- The unique field in the `where` option and the unique field in the `create` option have the same value
+
+If your query does not meet these criteria, then Prisma Client handles the upsert itself, and does not initiate a native upsert.
```
Comment (garrensmith) [inline content/200-concepts/100-components/02-prisma-client/260-native-upserts.mdx]: I would change this to:

If your query does not meet these criteria, then Prisma Client handles the upsert itself.

File: content/200-concepts/100-components/02-prisma-client/260-native-upserts.mdx [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+    userName: 'Alice',
+    email: 'alice@prisma.io',
+  },
+  update: {
+    email: 'updated@example.com',
+  },
+})
+```
+
+In the following query, the values for `userName` in the `where` and `create` options are different, so Prisma Client does **not** initiate a native upsert.
```
Comment (garrensmith) [inline content/200-concepts/100-components/02-prisma-client/260-native-upserts.mdx]: Awesome examples

**Review by garrensmith — APPROVED** [review]
Really nice. I've given some minor wording feedback but feel free to ignore that.

File: content/400-reference/200-api-reference/050-prisma-client-reference.mdx [hunk 5]
```diff
@@ -955,6 +955,7 @@ const user = await prisma.user.update({
 
 - To perform arithmetic operations on update (add, subtract, multiply, divide), use [atomic updates](#atomic-number-operations) to prevent race conditions.
 - If two or more upsert operations happen at the same time and the record doesn't already exist, then a race condition might happen. As a result, one or more of the upsert operations might throw a unique key constraint error. Your application code can catch this error and retry the operation. [Learn more](#unique-key-constraint-errors-on-upserts).
+- From version 4.6.0, Prisma carries out native upsers for Postgres, SQLite, and CockroachDB databases, if specific crieria are met. [Learn more](/concepts/components/prisma-client/native-upserts).
```
Comment (andrew-walford-prisma) [inline content/400-reference/200-api-reference/050-prisma-client-reference.mdx]:
```suggestion
- From version 4.6.0, Prisma carries out native upserts for Postgres, SQLite, and CockroachDB databases, if specific crieria are met. [Learn more](/concepts/components/prisma-client/native-upserts).
```

File: content/400-reference/200-api-reference/050-prisma-client-reference.mdx [hunk 6]
```diff
@@ -955,6 +955,7 @@ const user = await prisma.user.update({
 
 - To perform arithmetic operations on update (add, subtract, multiply, divide), use [atomic updates](#atomic-number-operations) to prevent race conditions.
 - If two or more upsert operations happen at the same time and the record doesn't already exist, then a race condition might happen. As a result, one or more of the upsert operations might throw a unique key constraint error. Your application code can catch this error and retry the operation. [Learn more](#unique-key-constraint-errors-on-upserts).
+- From version 4.6.0, Prisma carries out native upserts for Postgres, SQLite, and CockroachDB databases, if specific crieria are met. [Learn more](/concepts/components/prisma-client/native-upserts).
```
Comment (janpio) [inline content/400-reference/200-api-reference/050-prisma-client-reference.mdx]:
```suggestion
- From version 4.6.0, Prisma carries out native upserts for PostgreSQL, SQLite, and CockroachDB databases, if specific criteria are met. [Learn more](/concepts/components/prisma-client/native-upserts).
```

File: content/400-reference/200-api-reference/050-prisma-client-reference.mdx [hunk 7]
```diff
@@ -932,7 +932,7 @@ const user = await prisma.user.update({
 
 ### <inlinecode>upsert</inlinecode>
 
-`upsert` updates an existing _or_ creates a new database record.
+`upsert` updates an existing database record _or_ creates a new database record.
```
Comment (janpio) [inline content/400-reference/200-api-reference/050-prisma-client-reference.mdx]: Should we maybe add when it does what? 

Badly worded suggestion:
> ... depending if one for the `where` condition already exists.

File: content/200-concepts/100-components/02-prisma-client/260-native-upserts.mdx [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+    profileViews: 1,
+    userName: 'Alice',
+    email: 'alice@prisma.io',
+  },
+  update: {
+    email: 'updated@example.com',
+  },
+})
+```
+
```
Comment (janpio) [inline content/200-concepts/100-components/02-prisma-client/260-native-upserts.mdx]: Would be great if we could show the full generated and executed SQL here.

File: content/200-concepts/100-components/02-prisma-client/260-native-upserts.mdx [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
```
Comment (janpio) [inline content/200-concepts/100-components/02-prisma-client/260-native-upserts.mdx]: Would be great if we could show the full generated and executed SQL here.

File: content/200-concepts/100-components/02-prisma-client/260-native-upserts.mdx [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+- You use Prisma version 4.6.0 or later
+- Your application uses a CockroachDB, PostgreSQL, or SQLite database
+
+### <inlinecode>upsert</inlinecode> query criteria
+
+Prisma Client uses a native upsert if your `upsert` query meets the following criteria:
+
+- There are no nested queries in the `upsert`'s `create` and `update` [options](/reference/api-reference/prisma-client-reference/#options-7)
+- The query updates only one model
```
Comment (janpio) [inline content/200-concepts/100-components/02-prisma-client/260-native-upserts.mdx]: Is "update" the correct terminology here? Does this only apply to the `update` part? Or "affect"?

File: content/200-concepts/100-components/02-prisma-client/260-native-upserts.mdx [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+- There is only one unique field in the `upsert`'s `where` option
+- The unique field in the `where` option and the unique field in the `create` option have the same value
```
Comment (janpio) [inline content/200-concepts/100-components/02-prisma-client/260-native-upserts.mdx]: What happens if there is no `where` @garrensmith?

**Comment by janpio** [comment]
The page itself and the information contained in it are fine 👍 

What I am not sure about is having a standalone page "Native Upserts" as that is more or less an implementation detail of how Prisma's `upsert` works under the hood. For our users this in 99% of cases should not be relevant to know about. 

(Additionally "Native upsert" really means "database native upsert" which is not _really_ true, as `INSERT ON CONFLICT` is (pedantic nitpick incoming) not a _native upsert_, but an implementation of an upsert in PostgreSQL.)

So what can we do about this? 

Maybe we do a "Prisma Query Engine under the hood" page (or a similar, shorter name that makes clear, that its content explains some of the implementation details that we usually just brush away as internal) and then we have a "Prisma vs. database/atomic/native upsert" section. It would start by explaining that Prisma chooses the best SQL query possible for each case, and that for `upsert` that means depending on some criteria using the native database option, or going with something that gives almost the same result but by using different queries (which then is illustrated by the content the page currently has).

File: content/200-concepts/100-components/02-prisma-client/260-native-upserts.mdx (hunk 11 above)
Comment (garrensmith) [reply content/200-concepts/100-components/02-prisma-client/260-native-upserts.mdx]: It has to have a where clause for the engine to accept it.

**Comment by garrensmith** [comment]
@janpio My thinking with database level upserts (native upserts) or whatever we call them is that a user that understands that they exist will want to use it. So we do need to make it easy for them to find in the docs so they can make sure they get the best performance and behaviour from prisma. So an under the hood section seems fine as long we as point them to it in the upsert docs.

**Comment by janpio** [comment]
Oh yes, obviously this needs to be linked where we introduce `upsert` (reference and CRUD). There it should probably say that "when possible this uses a database native upsert SQL command, otherwise Prisma takes over" (or similar).

**Comment by andrew-walford-prisma** [comment]
As discussed in Slack, I'm going to pull this information back into the Client API for now. At a later date we'll create a "Prisma under the hood" page, and this information will be part of that. For terminology, I'll go with "database upsert" instead of "native upsert".

File: content/200-concepts/100-components/02-prisma-client/260-native-upserts.mdx (hunk 10 above)
Comment (andrew-walford-prisma) [reply content/200-concepts/100-components/02-prisma-client/260-native-upserts.mdx]: Good catch. I've updated this to "modifies"

File: content/200-concepts/100-components/02-prisma-client/260-native-upserts.mdx (hunk 8 above)
Comment (andrew-walford-prisma) [reply content/200-concepts/100-components/02-prisma-client/260-native-upserts.mdx]: Noted - I'll investigate

File: content/200-concepts/100-components/02-prisma-client/260-native-upserts.mdx (hunk 9 above)
Comment (andrew-walford-prisma) [reply content/200-concepts/100-components/02-prisma-client/260-native-upserts.mdx]: Noted - I'll investigate

File: content/200-concepts/100-components/02-prisma-client/260-native-upserts.mdx (hunk 8 above)
Comment (andrew-walford-prisma) [reply content/200-concepts/100-components/02-prisma-client/260-native-upserts.mdx]: I've added the SQL output for the example that results in a database upsert

File: content/400-reference/200-api-reference/050-prisma-client-reference.mdx [hunk 12]
```diff
@@ -1005,6 +1005,111 @@ When your application tries to perform two or more concurrent upsert operations,
 
 Handle the P2002 error in your application code. When it occurs, retry the upsert operation to update the row.
 
+#### Database upserts
+
+Where possible, Prisma Client hands over an `upsert` query to the database, instead of handling the upsert itself. This is called a _database upsert_.
+
+Database upserts have the following advantages:
+
+- They are faster than regular upserts
```
Comment (janpio) [inline content/400-reference/200-api-reference/050-prisma-client-reference.mdx]: "regular upserts" might be misleading here, the database ones are "regular" for most.
"Prisma level upserts" maybe?

File: content/400-reference/200-api-reference/050-prisma-client-reference.mdx [hunk 13]
```diff
@@ -1005,6 +1005,111 @@ When your application tries to perform two or more concurrent upsert operations,
...
+- [Unique key constraint errors](#unique-key-constraint-errors-on-upserts) cannot happen
+
+Prisma Client uses a database upsert automatically when [specific criteria](#criteria-for-database-upserts) are met.
```
Comment (janpio) [inline content/400-reference/200-api-reference/050-prisma-client-reference.mdx]: Does this link need updating?

File: content/400-reference/200-api-reference/050-prisma-client-reference.mdx [hunk 14]
```diff
@@ -1005,6 +1005,111 @@ When your application tries to perform two or more concurrent upsert operations,
...
+
+To use a database upsert, Prisma Client sends the SQL construction [`INSERT ... ON CONFLICT SET .. WHERE`](www.prisma.io/dataguide/postgresql/inserting-and-modifying-data/insert-on-conflict) to the database.
```
Comment (janpio) [inline content/400-reference/200-api-reference/050-prisma-client-reference.mdx]: https:// missing on link

**Review by janpio — APPROVED** [review]

File: content/400-reference/200-api-reference/050-prisma-client-reference.mdx (hunk 12 above)
Comment (andrew-walford-prisma) [reply content/400-reference/200-api-reference/050-prisma-client-reference.mdx]: Good point - I'll revise

File: content/400-reference/200-api-reference/050-prisma-client-reference.mdx (hunk 13 above)
Comment (andrew-walford-prisma) [reply content/400-reference/200-api-reference/050-prisma-client-reference.mdx]: Yep - I'll fix

File: content/400-reference/200-api-reference/050-prisma-client-reference.mdx (hunk 14 above)
Comment (andrew-walford-prisma) [reply content/400-reference/200-api-reference/050-prisma-client-reference.mdx]: Fixed!

File: content/400-reference/200-api-reference/050-prisma-client-reference.mdx (hunk 12 above)
Comment (andrew-walford-prisma) [reply content/400-reference/200-api-reference/050-prisma-client-reference.mdx]: Done

**Review by janpio — APPROVED** [review]

**Pull request merged**